src/data_device/mod.rs

```diff
@@ -0,0 +1,315 @@
+        );
+        $crate::reexports::client::delegate_dispatch!($ty:
+            [
+                $crate::reexports::client::protocol::wl_data_offer::WlDataOffer: $crate::data_device::data_offer::DataOfferData
+            ] => $crate::data_device::DataDeviceManagerState
+        );
+        $crate::reexports::client::delegate_dispatch!($ty:
+            [
+                $crate::reexports::client::protocol::wl_data_device::WlDataDevice: $crate::data_device::data_device::DataDeviceData
+            ] => $crate::data_device::DataDeviceManagerState
+        );
+    };
+}
+
+/// A file descriptor that can only be read from
+///
+/// If the `calloop` cargo feature is enabled, this can be used
+/// as an `EventSource` in a calloop event loop.
+#[derive(Debug)]
+pub struct ReadPipe {
+    #[cfg(feature = "calloop")]
+    file: calloop::generic::Generic<fs::File>,
+    #[cfg(not(feature = "calloop"))]
+    file: fs::File,
+}
+
+#[cfg(feature = "calloop")]
+impl io::Read for ReadPipe {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.file.file.read(buf)
+    }
+}
+
+#[cfg(not(feature = "calloop"))]
+impl io::Read for ReadPipe {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.file.read(buf)
+    }
+}
+
+#[cfg(feature = "calloop")]
+impl FromRawFd for ReadPipe {
+    unsafe fn from_raw_fd(fd: RawFd) -> ReadPipe {
+        ReadPipe {
+            file: calloop::generic::Generic::new(
+                unsafe { FromRawFd::from_raw_fd(fd) },
+                calloop::Interest::READ,
+                calloop::Mode::Level,
+            ),
+        }
+    }
+}
+
+#[cfg(not(feature = "calloop"))]
+impl FromRawFd for ReadPipe {
+    unsafe fn from_raw_fd(fd: RawFd) -> ReadPipe {
+        ReadPipe { file: unsafe { FromRawFd::from_raw_fd(fd) } }
+    }
+}
+
+#[cfg(feature = "calloop")]
+impl AsRawFd for ReadPipe {
+    fn as_raw_fd(&self) -> RawFd {
+        self.file.file.as_raw_fd()
+    }
+}
+
+#[cfg(not(feature = "calloop"))]
+impl AsRawFd for ReadPipe {
+    fn as_raw_fd(&self) -> RawFd {
+        self.file.as_raw_fd()
+    }
+}
+
+#[cfg(feature = "calloop")]
+impl IntoRawFd for ReadPipe {
+    fn into_raw_fd(self) -> RawFd {
+        self.file.file.into_raw_fd()
+    }
+}
+
+#[cfg(not(feature = "calloop"))]
+impl IntoRawFd for ReadPipe {
+    fn into_raw_fd(self) -> RawFd {
+        self.file.into_raw_fd()
+    }
+}
+
+#[cfg(feature = "calloop")]
+impl calloop::EventSource for ReadPipe {
+    type Event = ();
+    type Error = std::io::Error;
+    type Metadata = fs::File;
+    type Ret = ();
+
+    fn process_events<F>(
+        &mut self,
+        readiness: calloop::Readiness,
+        token: calloop::Token,
+        mut callback: F,
+    ) -> std::io::Result<calloop::PostAction>
+    where
+        F: FnMut((), &mut fs::File),
+    {
+        self.file.process_events(readiness, token, |_, file| {
+            callback((), file);
+            Ok(calloop::PostAction::Continue)
```

Comment:
I believe it is not meant to be a oneshot, but I think that was just an easy way for me to use it in the example. A user really probably shouldn't try to read all the data from the pipe at once, considering it might take a while.

Reply:
If that is the case, then how would you know all the data has been copied if readiness could lie?